common/codex32.c

```diff
@@ -0,0 +1,170 @@
+#include "config.h"
+#include <assert.h>
+#include <bitcoin/chainparams.h>
+#include <ccan/tal/str/str.h>
+#include <common/bech32.h>
+#include <common/bech32_util.h>
+#include <common/bolt12.h>
+#include <common/bolt12_merkle.h>
+#include <common/codex32.h>
+#include <common/configdir.h>
+#include <common/features.h>
+#include <math.h>
+#include <secp256k1_schnorrsig.h>
+#include <time.h>
+
+
+/* Helper to sanity check the codex32 string parts */
+static bool sanity_check (struct codex32_parts *parts,
+			  char **fail)
+{
+	if (!streq(parts->hrp, "ms") && !streq(parts->hrp, "MS")) {
+		*fail = tal_fmt(tmpctx, "Invalid HRP!");
+		return false;
+	}
+	if (parts->threshold > 9 ||
+	    parts->threshold < 0 ||
+	    parts->threshold == 1) {
+		*fail = tal_fmt(tmpctx, "Invalid threshold!");
+		return false;
+	}
+	if (strlen(parts->id) != 4) {
+		*fail = tal_fmt(tmpctx, "Invalid ID!");
+		return false;
+	}
+	if ((parts->threshold == 0 && !(*(parts->share_idx) == 'S' || *(parts->share_idx) == 's')))
+	{
+		*fail = tal_fmt(tmpctx, "Expected share index S for threshold 0!");
+		return false;
+	}
+	if((strlen(parts->payload) * 5) % 8 > 4) {
+		*fail = tal_fmt(tmpctx, "Incomplete group exist in payload!");
+		return false;
+	}
+
+	return true;
+}
+
+static const u8 *extract_data(struct codex32_parts *parts)
+{
+	// FIXME: Make sure the size of array is correct, because the documentation has 1 extra byte...
+	u8 *ret = tal_arr(parts, u8, ((strlen(parts->payload) * 5 + 7) / 8) - 1);
+
+	uint8_t next_byte = 0;
+	uint8_t rem = 0;
+	size_t i = 0, j = 0;
+	while (parts->payload[i] != '\0') {
+		char ch = parts->payload[i++];
+		uint8_t fe = bech32_charset_rev[(int)ch];
+
+		if (rem < 3) {
+			// If we are within 3 bits of the start we can fit the whole next char in
+			next_byte |= fe << (3 - rem);
+		}
+		else if (rem == 3) {
+			// If we are exactly 3 bits from the start then this char fills in the byte
+			ret[j++] = next_byte | fe;
+			next_byte = 0;
+		}
+		else { // rem > 3
+			// Otherwise we have to break it in two
+			uint8_t overshoot = rem - 3;
+			assert(overshoot > 0);
+			ret[j++] = next_byte | (fe >> overshoot);
+			next_byte = fe << (8 - overshoot);
+		}
+
+		rem = (rem + 5) % 8;
+	}
+	assert(rem <= 4); // checked when parsing the string
+	return ret;
+}
+
+static bool case_check(const char *codex32str)
+{
+	bool have_lower = false, have_upper = false;
+	size_t str_len = strlen(codex32str);
+	for (size_t i = 0; i < str_len; i++) {
+		if (codex32str[i] >= 'a' && codex32str[i] <= 'z') {
+            		have_lower = true;
+		} else if (codex32str[i] >= 'A' && codex32str[i] <= 'Z') {
+			have_upper = true;
+		}
+	}
+	if (have_lower && have_upper) {
+		return false;
+	}
+	return true;
+}
+
+/* Return `False` if the codex32 is invalid */
+bool codex32_decode(const char *codex32str,
+		    struct codex32_parts *parts,
+		    const u8 **payload,
+		    char **fail)
```

Comment:
again, use "returns failmsg or NULL" pattern.

Reply:
Actually, it should return the struct codex32_parts (why not just call that, `struct codex32`?):

```
struct codex32 * codex32_decode(const tal_t *ctx, const char *codex32str, char **fail)
```

And expose `extract_data` as:

```
const char *codex32_encode(const tal_t *ctx, const struct codex32 *codex32)
```
?